File: macos/Sources/VPSQuotaCore/Core/BillingPeriod.swift
import Foundation

/// 一个账单周期，左闭右开：`[start, end)`。
///
/// 服务商的流量额度按账单日重置，而账单日不一定是 1 号
/// （DMIT 通常是开通日，Vultr 是账号计费日），所以不能简单地按自然月统计。
public struct BillingPeriod: Hashable, Sendable {
    /// 账期起点（UTC 零点，含）
    public let start: Date
    /// 账期终点（UTC 零点，不含）
    public let end: Date

    public init(start: Date, end: Date) {
        self.start = start
        self.end = end
    }

    /// 起点的 `YYYY-MM-DD`，可直接用于 SQLite 的 `day >= ?`。
    public var startDay: String { UTCDay.string(from: start) }
    /// 终点的 `YYYY-MM-DD`（不含），可直接用于 SQLite 的 `day < ?`。
    public var endDayExclusive: String { UTCDay.string(from: end) }

    /// 账期总天数。
    public var totalDays: Int {
        UTCDay.calendar.dateComponents([.day], from: start, to: end).day ?? 30
    }

    /// 账期已过去的天数（含小数）。超出账期时钳制到 `[0, totalDays]`。
    public func elapsedDays(now: Date) -> Double {
        let seconds = now.timeIntervalSince(start)
        let days = seconds / 86_400
        return Swift.min(Swift.max(days, 0), Double(totalDays))
    }

    /// 距账期结束还剩多少个完整日历天。
    public func remainingDays(now: Date) -> Int {
        let days = UTCDay.calendar.dateComponents(
            [.day], from: UTCDay.startOfDay(now), to: end
        ).day ?? 0
        return Swift.max(0, days)
    }

    /// 判断某个 `YYYY-MM-DD` 是否落在本账期内。
    public func contains(day: String) -> Bool {
        day >= startDay && day < endDayExclusive
    }

    /// 求出包含 `now` 的账期。
    ///
    /// `resetDay` 会被钳制到 1…31；当月天数不足时（例如 31 号遇上 2 月）
    /// 取当月最后一天，这与主流服务商的做法一致。
    public static func current(
        resetDay: Int,
        now: Date = Date()
    ) -> BillingPeriod {
        let cal = UTCDay.calendar
        let clamped = Swift.min(Swift.max(resetDay, 1), 31)

        // 本月的重置时刻
        let thisMonthAnchor = anchor(inMonthOf: now, resetDay: clamped)

        if now >= thisMonthAnchor {
            // 已过本月重置日：账期是 [本月重置日, 下月重置日)
            let nextMonth = cal.date(byAdding: .month, value: 1, to: thisMonthAnchor)!
            return BillingPeriod(
                start: thisMonthAnchor,
                end: anchor(inMonthOf: nextMonth, resetDay: clamped)
            )
        } else {
            // 尚未到本月重置日：账期是 [上月重置日, 本月重置日)
            let prevMonth = cal.date(byAdding: .month, value: -1, to: thisMonthAnchor)!
            return BillingPeriod(
                start: anchor(inMonthOf: prevMonth, resetDay: clamped),
                end: thisMonthAnchor
            )
        }
    }

    /// 求 `date` 所在自然月中的重置时刻（UTC 零点）。
    ///
    /// 当月天数不足 `resetDay` 时退化为当月最后一天 —— 这是本函数存在的全部理由，
    /// 也是 `resetDay = 29/30/31` 时唯一的正确处理方式。
    private static func anchor(inMonthOf date: Date, resetDay: Int) -> Date {
        let cal = UTCDay.calendar
        let comps = cal.dateComponents([.year, .month], from: date)
        let firstOfMonth = cal.date(from: comps)!
        let daysInMonth = cal.range(of: .day, in: .month, for: firstOfMonth)!.count
        let day = Swift.min(resetDay, daysInMonth)

        var target = comps
        target.day = day
        target.hour = 0
        target.minute = 0
        target.second = 0
        return cal.date(from: target)!
    }
}

File: windows/VpsQuota/Core/BillingPeriod.cs
namespace VpsQuota.Core;

/// <summary>
/// 一个账单周期，左闭右开：<c>[Start, End)</c>。
///
/// 服务商的流量额度按账单日重置，而账单日不一定是 1 号
/// （DMIT 通常是开通日，Vultr 是账号计费日），所以不能简单地按自然月统计。
/// </summary>
public readonly record struct BillingPeriod(DateTime Start, DateTime End)
{
    /// <summary>起点的 yyyy-MM-dd，可直接用于 SQLite 的 <c>day &gt;= ?</c>。</summary>
    public string StartDay => UtcDay.String(Start);

    /// <summary>终点的 yyyy-MM-dd（不含），可直接用于 SQLite 的 <c>day &lt; ?</c>。</summary>
    public string EndDayExclusive => UtcDay.String(End);

    public int TotalDays => (int)Math.Round((End - Start).TotalDays);

    /// <summary>账期已过去的天数（含小数）。超出账期时钳制到 [0, TotalDays]。</summary>
    public double ElapsedDays(DateTime now) =>
        Math.Clamp((ToUtc(now) - Start).TotalDays, 0, TotalDays);

    /// <summary>距账期结束还剩多少个完整日历天。</summary>
    public int RemainingDays(DateTime now) =>
        Math.Max(0, (int)Math.Round((End - UtcDay.StartOfDay(ToUtc(now))).TotalDays));

    /// <summary>判断某个 yyyy-MM-dd 是否落在本账期内。</summary>
    public bool Contains(string day) =>
        string.CompareOrdinal(day, StartDay) >= 0 &&
        string.CompareOrdinal(day, EndDayExclusive) < 0;

    /// <summary>
    /// 求出包含 <paramref name="now"/> 的账期。
    ///
    /// <paramref name="resetDay"/> 会被钳制到 1…31；当月天数不足时
    /// （例如 31 号遇上 2 月）取当月最后一天，这与主流服务商的做法一致。
    /// </summary>
    public static BillingPeriod Current(int resetDay, DateTime now)
    {
        // 账期一律按 UTC 切分（macOS 端由 UTC Calendar 强制保证）。
        // 这里传进来的必须是 UTC 时刻：本地时刻会让 UTC+8 的机器在当地 00:00–08:00
        // 把 now 算成前一天；若那天正好是重置日，now >= anchor 的判定翻转，
        // 整个账期回退一个月，已用量直接错一整期。类型系统拦不住，只能显式转。
        now = ToUtc(now);

        var clamped = Math.Clamp(resetDay, 1, 31);
        var thisMonthAnchor = Anchor(now, clamped);

        if (now >= thisMonthAnchor)
        {
            // 已过本月重置日：账期是 [本月重置日, 下月重置日)
            return new BillingPeriod(thisMonthAnchor, Anchor(thisMonthAnchor.AddMonths(1), clamped));
        }

        // 尚未到本月重置日：账期是 [上月重置日, 本月重置日)
        return new BillingPeriod(Anchor(thisMonthAnchor.AddMonths(-1), clamped), thisMonthAnchor);
    }

    /// <summary>
    /// 求 <paramref name="date"/> 所在自然月中的重置时刻（UTC 零点）。
    ///
    /// 当月天数不足 <paramref name="resetDay"/> 时退化为当月最后一天 ——
    /// 这是本函数存在的全部理由，也是 resetDay = 29/30/31 时唯一的正确处理方式。
    /// </summary>
    /// <summary>
    /// 统一成 UTC 时刻。Local 按时区换算；Unspecified 视作已经是 UTC
    /// （应用内的时间全部来自 DateTime.UtcNow 或本文件构造的 UTC 值）。
    /// </summary>
    private static DateTime ToUtc(DateTime value) => value.Kind switch
    {
        DateTimeKind.Utc => value,
        DateTimeKind.Local => value.ToUniversalTime(),
        _ => DateTime.SpecifyKind(value, DateTimeKind.Utc),
    };

    private static DateTime Anchor(DateTime date, int resetDay)
    {
        var daysInMonth = DateTime.DaysInMonth(date.Year, date.Month);
        var day = Math.Min(resetDay, daysInMonth);
        return new DateTime(date.Year, date.Month, day, 0, 0, 0, DateTimeKind.Utc);
    }
}
